sdk/appconfiguration/feature-management/feature/management/_defaultfilters.py

```diff
@@ -0,0 +1,112 @@
+# ------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# -------------------------------------------------------------------------
+from ._featurefilters import FeatureFilter
+from datetime import datetime, timezone
+from email.utils import parsedate_to_datetime
+import logging
+import hashlib
+
+
+class TargetingException(Exception):
+    pass
+
+
+class TimeWindowFilter(FeatureFilter):
+    def evaluate(self, context, **kwargs):
+        """Determain if the feature flag is enabled for the given context"""
+        start = context.get("parameters", {}).get("Start")
+        end = context.get("parameters", {}).get("End")
+
+        current_time = datetime.now(timezone.utc)
+
+        if not start and not end:
+            logging.warn("TimeWindowFilter: Either Start or End are required to have as a parameter")
+            return False
+
+        start_time = parsedate_to_datetime(start) if start else None
+        end_time = parsedate_to_datetime(end) if end else None
+
+        return (start_time is None or start_time <= current_time) and (end_time is None or current_time < end_time)
+
+
+class TargetingFilter(FeatureFilter):
+    @staticmethod
+    def _is_targeted(contextId, rollout_percentage):
+        """Determine if the user is targeted for the given context"""
+        # Alway return true if rollout percentage is 100
+        if rollout_percentage == 100:
+            return True
+
+        hashed_contextId = hashlib.sha256(contextId.encode()).hexdigest()
+        contextMarker = abs(int(hashed_contextId, 16))
+        percentage = (contextMarker / (2**256 - 1)) * 100
+
+        return percentage < rollout_percentage
+
+    def _target_group(self, target_user, target_group, group, feature_flag_name):
+        group_rollout_percentage = group.get("RolloutPercentage", 0)
+        audienceContextId = target_user + "\n" + feature_flag_name + "\n" + group.get("Name", "")
+
+        return self._is_targeted(audienceContextId, group_rollout_percentage)
+
+    def evaluate(self, context, **kwargs):
+        """Determain if the feature flag is enabled for the given context"""
+        target_user = kwargs.pop("user", None)
+        target_groups = kwargs.pop("groups", [])
+        ignore_case = kwargs.pop("ignore_case", False)
+
+        if not target_user and not (target_groups and len(target_groups) > 0):
+            logging.warn("TargetingFilter: Name or Groups are required parameters")
+            return False
+
+        audience = context.get("parameters", {}).get("Audience", None)
+        feature_flag_name = context.get("name", None)
+        if not audience:
+            raise TargetingException("Audience is required for TargetingFilter")
```

Comment:
Same here and in all error messages, I think its better to use class/attribute name